pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1367 @@
+					.checked_add(&payment).ok_or(Error::<T>::Overflow)?;
+				Oracles::<T>::insert(&oracle, oracle_meta);
+
+				// delete the details if the maximum count has been reached
+				if details.submissions.len() < max_count as usize {
+					Details::<T>::insert(feed_id, round_id, details);
+				}
+
+				Ok(().into())
+			})
+		}
+
+		/// Disable and add oracles for the given feed.
+		/// Limited to the owner of a feed.
+		#[weight = T::WeightInfo::change_oracles(to_disable.len() as u32, to_add.len() as u32)]
+		pub fn change_oracles(
+			origin,
+			feed_id: T::FeedId,
+			to_disable: Vec<T::AccountId>,
+			to_add: Vec<(T::AccountId, T::AccountId)>,
+		) -> DispatchResultWithPostInfo {
+			let owner = ensure_signed(origin)?;
+
+			let mut to_disable = to_disable;
+			to_disable.sort_unstable();
+			to_disable.dedup();
+			with_transaction_result(|| -> DispatchResultWithPostInfo {
+				// synced on drop
+				let mut feed = Feed::<T>::load_from(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+				feed.ensure_owner(&owner)?;
+				feed.disable_oracles(to_disable)?;
+				feed.add_oracles(to_add)?;
+
+				Ok(().into())
+			})
+		}
+
+		/// Update the configuration for future oracle rounds.
+		/// Limited to the owner of a feed.
+		#[weight = T::WeightInfo::update_future_rounds()]
+		pub fn update_future_rounds(
+			origin,
+			feed_id: T::FeedId,
+			payment: BalanceOf<T>,
+			submission_count_bounds: (u32, u32),
+			restart_delay: RoundId,
+			timeout: T::BlockNumber,
+		) -> DispatchResultWithPostInfo {
+			let owner = ensure_signed(origin)?;
+			// synced on drop
+			let mut feed = Feed::<T>::load_from(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+			feed.ensure_owner(&owner)?;
+
+			feed.update_future_rounds(payment, submission_count_bounds, restart_delay, timeout)?;
+
+			Ok(().into())
+		}
+
+		/// Prune the state of a feed to reduce storage load.
+		///
+		/// - Will update the `first_valid_round` to the most recent round kept.
+		/// - Will only prune until hitting the pruning window (which makes sure to keep N rounds
+		/// of data available).
+		///
+		/// Limited to the owner of a feed.
+		#[weight = T::WeightInfo::prune(keep_round.saturating_sub(*first_to_prune))]
+		pub fn prune(
+			origin,
+			feed_id: T::FeedId,
+			first_to_prune: RoundId,
+			keep_round: RoundId,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+			ensure!(first_to_prune > Zero::zero(), Error::<T>::CannotPruneRoundZero);
+			ensure!(keep_round > first_to_prune, Error::<T>::NothingToPrune);
+			let mut feed = Self::feed_config(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+			ensure!(feed.owner == owner, Error::<T>::NotFeedOwner);
+			let first_valid_round = feed.first_valid_round.ok_or(Error::<T>::NoValidRoundYet)?;
+			ensure!(first_to_prune <= first_valid_round, Error::<T>::PruneContiguously);
+			let pruning_window = T::PruningWindow::get();
+			ensure!(feed.latest_round.saturating_sub(first_to_prune) > pruning_window, Error::<T>::NothingToPrune);
```

Comment:
Do you need `first_to_prune`?, maybe you can always prune from `first_valid_round`? It seems to be a case where there is nothing to prune but the error is not triggered (like executing two times the test case?).

Reply:
Yes, I need it for weight estimation and because there might be rounds before the first valid round that also need to be cleaned up. (Namely rounds that just contain default data because the oracles lagged behind the start of the feed.)